header/sync/sync.go

```diff
@@ -119,138 +132,186 @@ func (s *Syncer) State() State {
+		}
 	}
 
-	// 3. Filter out maybeHead if behind trusted
-	if maybeHead.Height <= trstHead.Height {
-		log.Warnw("received known header",
-			"height", maybeHead.Height,
-			"hash", maybeHead.Hash())
-		return pubsub.ValidationIgnore // we don't know if header is invalid so ignore
-	}
+	// and if valid, set it as new head
+	s.pending.Add(maybeHead)
+	s.wantSync()
+	log.Infow("new pending head", "height", maybeHead.Height, "hash", maybeHead.Hash())
+	return pubsub.ValidationAccept
+}
 
-	// 4. Verify maybeHead against trusted
-	err = trstHead.VerifyNonAdjacent(maybeHead)
+// validate checks validity of given the header against the subjective head.
+func (s *Syncer) validate(ctx context.Context, new *header.ExtendedHeader) pubsub.ValidationResult {
+	sbjHead, err := s.subjectiveHead(ctx)
+	if err != nil {
+		log.Errorw("getting subjective head", "err", err)
+		return pubsub.ValidationIgnore // local error, so ignore
+	}
+	// if header is from the past, we don't know, so ignore
+	if !sbjHead.IsBefore(new) {
+		log.Warnw("received past header",
+			"height", new.Height,
+			"hash", new.Hash())
+		return pubsub.ValidationIgnore
+	}
+	// perform verification
+	err = sbjHead.VerifyNonAdjacent(new)
 	var verErr *header.VerifyError
 	if errors.As(err, &verErr) {
 		log.Errorw("invalid header",
-			"height_of_invalid", maybeHead.Height,
-			"hash_of_invalid", maybeHead.Hash(),
-			"height_of_trusted", trstHead.Height,
-			"hash_of_trusted", trstHead.Hash(),
+			"height_of_invalid", new.Height,
+			"hash_of_invalid", new.Hash(),
+			"height_of_subjective", sbjHead.Height,
+			"hash_of_subjective", sbjHead.Hash(),
 			"reason", verErr.Reason)
 		return pubsub.ValidationReject
 	}
-
-	// 5. Save verified header to pending cache
-	// NOTE: Pending cache can't be DOSed as we verify above each header against a trusted one.
-	s.pending.Add(maybeHead)
-	// and trigger sync to catch-up
-	s.wantSync()
-	log.Infow("pending head",
-		"height", maybeHead.Height,
-		"hash", maybeHead.Hash())
+	// and accept if the header is good
 	return pubsub.ValidationAccept
 }
 
-// syncTo requests headers from locally stored head up to the new head.
-func (s *Syncer) syncTo(ctx context.Context, newHead *header.ExtendedHeader) {
+// wantSync will trigger the syncing loop (non-blocking).
+func (s *Syncer) wantSync() {
+	select {
+	case s.triggerSync <- struct{}{}:
+	default:
+	}
+}
+
+// syncLoop controls syncing process.
+func (s *Syncer) syncLoop() {
+	for {
+		select {
+		case <-s.triggerSync:
+			s.sync(s.ctx)
+		case <-s.ctx.Done():
+			return
+		}
+	}
+}
+
+// sync ensures we are synced from the local header Store's Head up to the new subjective head
+func (s *Syncer) sync(ctx context.Context) {
+	newHead := s.pending.Head()
+	if newHead == nil {
+		return
+	}
+
 	head, err := s.store.Head(ctx)
 	if err != nil {
 		log.Errorw("getting head during sync", "err", err)
 		return
 	}
 
 	if head.Height == newHead.Height {
-		return
+		return // should never happen, but just in case
```

Comment:
If we are trying to catch "should never happen" case down here, maybe we can also check the case

`if head.Height >= newHead.Height`, log out a WARN and return?

Reply:
Done